File: packages/camera/camera_avfoundation/example/ios/RunnerTests/CameraFocusTests.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@import camera_avfoundation;
#if __has_include(<camera_avfoundation/camera_avfoundation-umbrella.h>)
@import camera_avfoundation.Test;
#endif
@import XCTest;
@import AVFoundation;

#import "CameraTestUtils.h"
#import "MockCaptureDevice.h"
#import "MockDeviceOrientationProvider.h"

@interface CameraFocusTests : XCTestCase
@property(readonly, nonatomic) FLTCam *camera;
@property(readonly, nonatomic) MockCaptureDevice *mockDevice;
@property(readonly, nonatomic) MockDeviceOrientationProvider *mockDeviceOrientationProvider;
@end

@implementation CameraFocusTests

- (void)setUp {
  MockCaptureDevice *mockDevice = [[MockCaptureDevice alloc] init];
  _mockDevice = mockDevice;
  _mockDeviceOrientationProvider = [[MockDeviceOrientationProvider alloc] init];

  _camera = FLTCreateCamWithCaptureSessionQueueAndMediaSettings(
      nil, nil, nil,
      ^NSObject<FLTCaptureDevice> *(void) {
        return mockDevice;
      },
      _mockDeviceOrientationProvider);
}

- (void)testAutoFocusWithContinuousModeSupported_ShouldSetContinuousAutoFocus {
  // AVCaptureFocusModeContinuousAutoFocus and AVCaptureFocusModeContinuousAutoFocus are supported
  _mockDevice.isFocusModeSupportedStub = ^BOOL(AVCaptureFocusMode mode) {
    return mode == AVCaptureFocusModeContinuousAutoFocus || mode == AVCaptureFocusModeAutoFocus;
  };

  __block BOOL setFocusModeContinuousAutoFocusCalled = NO;

  _mockDevice.setFocusModeStub = ^(AVCaptureFocusMode mode) {
    // Don't expect setFocusMode:AVCaptureFocusModeAutoFocus
    if (mode == AVCaptureFocusModeAutoFocus) {
      XCTFail(@"Unexpected call to setFocusMode");
    } else if (mode == AVCaptureFocusModeContinuousAutoFocus) {
      setFocusModeContinuousAutoFocusCalled = YES;
    }
  };

  // Run test
  [_camera applyFocusMode:FCPPlatformFocusModeAuto onDevice:_mockDevice];

  // Expect setFocusMode:AVCaptureFocusModeContinuousAutoFocus
  XCTAssertTrue(setFocusModeContinuousAutoFocusCalled);
}

- (void)testAutoFocusWithContinuousModeNotSupported_ShouldSetAutoFocus {
  // AVCaptureFocusModeContinuousAutoFocus is not supported
  // AVCaptureFocusModeAutoFocus is supported
  _mockDevice.isFocusModeSupportedStub = ^BOOL(AVCaptureFocusMode mode) {
    return mode == AVCaptureFocusModeAutoFocus;
  };

  __block BOOL setFocusModeAutoFocusCalled = NO;

  // Don't expect setFocusMode:AVCaptureFocusModeContinuousAutoFocus
  _mockDevice.setFocusModeStub = ^(AVCaptureFocusMode mode) {
    if (mode == AVCaptureFocusModeContinuousAutoFocus) {
      XCTFail(@"Unexpected call to setFocusMode");
    } else if (mode == AVCaptureFocusModeAutoFocus) {
      setFocusModeAutoFocusCalled = YES;
    }
  };

  // Run test
  [_camera applyFocusMode:FCPPlatformFocusModeAuto onDevice:_mockDevice];

  // Expect setFocusMode:AVCaptureFocusModeAutoFocus
  XCTAssertTrue(setFocusModeAutoFocusCalled);
}

- (void)testAutoFocusWithNoModeSupported_ShouldSetNothing {
  // No modes are supported
  _mockDevice.isFocusModeSupportedStub = ^BOOL(AVCaptureFocusMode mode) {
    return NO;
  };

  // Don't expect any setFocus
  _mockDevice.setFocusModeStub = ^(AVCaptureFocusMode mode) {
    XCTFail(@"Unexpected call to setFocusMode");
  };

  // Run test
  [_camera applyFocusMode:FCPPlatformFocusModeAuto onDevice:_mockDevice];
}

- (void)testLockedFocusWithModeSupported_ShouldSetModeAutoFocus {
  // AVCaptureFocusModeContinuousAutoFocus and AVCaptureFocusModeAutoFocus are supported
  _mockDevice.isFocusModeSupportedStub = ^BOOL(AVCaptureFocusMode mode) {
    return mode == AVCaptureFocusModeContinuousAutoFocus || mode == AVCaptureFocusModeAutoFocus;
  };

  __block BOOL setFocusModeAutoFocusCalled = NO;

  // Expect only setFocusMode:AVCaptureFocusModeAutoFocus
  _mockDevice.setFocusModeStub = ^(AVCaptureFocusMode mode) {
    if (mode == AVCaptureFocusModeContinuousAutoFocus) {
      XCTFail(@"Unexpected call to setFocusMode");
    } else if (mode == AVCaptureFocusModeAutoFocus) {
      setFocusModeAutoFocusCalled = YES;
    }
  };

  // Run test
  [_camera applyFocusMode:FCPPlatformFocusModeLocked onDevice:_mockDevice];

  XCTAssertTrue(setFocusModeAutoFocusCalled);
}

- (void)testLockedFocusWithModeNotSupported_ShouldSetNothing {
  _mockDevice.isFocusModeSupportedStub = ^BOOL(AVCaptureFocusMode mode) {
    return mode == AVCaptureFocusModeContinuousAutoFocus;
  };

  // Don't expect any setFocus
  _mockDevice.setFocusModeStub = ^(AVCaptureFocusMode mode) {
    XCTFail(@"Unexpected call to setFocusMode");
  };

  // Run test
  [_camera applyFocusMode:FCPPlatformFocusModeLocked onDevice:_mockDevice];
}

- (void)testSetFocusPointWithResult_SetsFocusPointOfInterest {
  // UI is currently in landscape left orientation
  _mockDeviceOrientationProvider.orientation = UIDeviceOrientationLandscapeLeft;
  // Focus point of interest is supported
  _mockDevice.focusPointOfInterestSupported = YES;

  __block BOOL setFocusPointOfInterestCalled = NO;
  _mockDevice.setFocusPointOfInterestStub = ^(CGPoint point) {
    if (point.x == 1 && point.y == 1) {
      setFocusPointOfInterestCalled = YES;
    }
  };

  // Run test
  [_camera setFocusPoint:[FCPPlatformPoint makeWithX:1 y:1]
          withCompletion:^(FlutterError *_Nullable error){
          }];

  // Verify the focus point of interest has been set
  XCTAssertTrue(setFocusPointOfInterestCalled);
}

- (void)testSetFocusPoint_WhenNotSupported_ReturnsError {
  // UI is currently in landscape left orientation
  _mockDeviceOrientationProvider.orientation = UIDeviceOrientationLandscapeLeft;
  // Exposure point of interest is not supported
  _mockDevice.focusPointOfInterestSupported = NO;

  XCTestExpectation *expectation = [self expectationWithDescription:@"Completion with error"];

  // Run
  [_camera setFocusPoint:[FCPPlatformPoint makeWithX:1 y:1]
          withCompletion:^(FlutterError *_Nullable error) {
            XCTAssertNotNil(error);
            XCTAssertEqualObjects(error.code, @"setFocusPointFailed");
            XCTAssertEqualObjects(error.message, @"Device does not have focus point capabilities");
            [expectation fulfill];
          }];

  // Verify
  [self waitForExpectationsWithTimeout:1 handler:nil];
}

@end

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/CameraFocusTests.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import AVFoundation
import XCTest

@testable import camera_avfoundation

final class CameraFocusTests: XCTestCase {
  private func createSutAndMocks() -> (FLTCam, MockCaptureDevice, MockDeviceOrientationProvider) {
    let mockDevice = MockCaptureDevice()
    let mockDeviceOrientationProvider = MockDeviceOrientationProvider()

    let configuration = FLTCreateTestCameraConfiguration()
    configuration.captureDeviceFactory = { mockDevice }
    configuration.deviceOrientationProvider = mockDeviceOrientationProvider
    let camera = FLTCreateCamWithConfiguration(configuration)

    return (camera, mockDevice, mockDeviceOrientationProvider)
  }

  func testAutoFocusWithContinuousModeSupported_ShouldSetContinuousAutoFocus() {
    let (camera, mockDevice, _) = createSutAndMocks()
    // AVCaptureFocusModeContinuousAutoFocus and AVCaptureFocusModeAutoFocus are supported.
    mockDevice.isFocusModeSupportedStub = { mode in
      mode == .continuousAutoFocus || mode == .autoFocus
    }

    var setFocusModeContinuousAutoFocusCalled = false

    mockDevice.setFocusModeStub = { mode in
      // Don't expect setFocusMode:AVCaptureFocusModeAutoFocus.
      if mode == .autoFocus {
        XCTFail("Unexpected call to setFocusMode")
      } else if mode == .continuousAutoFocus {
        setFocusModeContinuousAutoFocusCalled = true
      }
    }

    camera.applyFocusMode(.auto, on: mockDevice)

    XCTAssertTrue(setFocusModeContinuousAutoFocusCalled)
  }

  func testAutoFocusWithContinuousModeNotSupported_ShouldSetAutoFocus() {
    let (camera, mockDevice, _) = createSutAndMocks()
    // AVCaptureFocusModeContinuousAutoFocus is not supported.
    // AVCaptureFocusModeAutoFocus is supported.
    mockDevice.isFocusModeSupportedStub = { mode in
      mode == .autoFocus
    }

    var setFocusModeAutoFocusCalled = false

    // Don't expect setFocusMode:AVCaptureFocusModeContinuousAutoFocus.
    mockDevice.setFocusModeStub = { mode in
      if mode == .continuousAutoFocus {
        XCTFail("Unexpected call to setFocusMode")
      } else if mode == .autoFocus {
        setFocusModeAutoFocusCalled = true
      }
    }

    camera.applyFocusMode(.auto, on: mockDevice)

    XCTAssertTrue(setFocusModeAutoFocusCalled)
  }

  func testAutoFocusWithNoModeSupported_ShouldSetNothing() {
    let (camera, mockDevice, _) = createSutAndMocks()
    // No modes are supported.
    mockDevice.isFocusModeSupportedStub = { _ in
      false
    }

    // Don't expect any setFocus.
    mockDevice.setFocusModeStub = {
      _ in XCTFail("Unexpected call to setFocusMode")
    }

    camera.applyFocusMode(.auto, on: mockDevice)
  }

  func testLockedFocusWithModeSupported_ShouldSetModeAutoFocus() {
    let (camera, mockDevice, _) = createSutAndMocks()
    // AVCaptureFocusModeContinuousAutoFocus and AVCaptureFocusModeAutoFocus are supported.
    mockDevice.isFocusModeSupportedStub = { mode in
      mode == .continuousAutoFocus || mode == .autoFocus
    }

    var setFocusModeAutoFocusCalled = false

    // Expect only setFocusMode:AVCaptureFocusModeAutoFocus.
    mockDevice.setFocusModeStub = { mode in
      if mode == .continuousAutoFocus {
        XCTFail("Unexpected call to setFocusMode")
      } else if mode == .autoFocus {
        setFocusModeAutoFocusCalled = true
      }
    }

    camera.applyFocusMode(.locked, on: mockDevice)

    XCTAssertTrue(setFocusModeAutoFocusCalled)
  }

  func testLockedFocusWithModeNotSupported_ShouldSetNothing() {
    let (camera, mockDevice, _) = createSutAndMocks()
    mockDevice.isFocusModeSupportedStub = { mode in
      mode == .continuousAutoFocus
    }

    // Don't expect any setFocus.
    mockDevice.setFocusModeStub = { _ in
      XCTFail("Unexpected call to setFocusMode")
    }

    camera.applyFocusMode(.locked, on: mockDevice)
  }

  func testSetFocusPointWithResult_SetsFocusPointOfInterest() {
    let (camera, mockDevice, mockDeviceOrientationProvider) = createSutAndMocks()
    // UI is currently in landscape left orientation.
    mockDeviceOrientationProvider.orientation = .landscapeLeft
    // Focus point of interest is supported.
    mockDevice.focusPointOfInterestSupported = true

    var setFocusPointOfInterestCalled = false
    mockDevice.setFocusPointOfInterestStub = { point in
      if point == CGPoint(x: 1.0, y: 1.0) {
        setFocusPointOfInterestCalled = true
      }
    }

    camera.setFocusPoint(FCPPlatformPoint.makeWith(x: 1, y: 1)) { error in
      XCTAssertNil(error)
    }

    XCTAssertTrue(setFocusPointOfInterestCalled)
  }

  func testSetFocusPoint_WhenNotSupported_ReturnsError() {
    let (camera, mockDevice, mockDeviceOrientationProvider) = createSutAndMocks()
    // UI is currently in landscape left orientation.
    mockDeviceOrientationProvider.orientation = .landscapeLeft
    // Focus point of interest is not supported.
    mockDevice.focusPointOfInterestSupported = false

    let expectation = self.expectation(description: "Completion with error")

    camera.setFocusPoint(FCPPlatformPoint.makeWith(x: 1, y: 1)) { error in
      XCTAssertNotNil(error)
      XCTAssertEqual(error?.code, "setFocusPointFailed")
      XCTAssertEqual(error?.message, "Device does not have focus point capabilities")
      expectation.fulfill()
    }

    waitForExpectations(timeout: 30, handler: nil)
  }
}
